### PR TITLE
Add support for batch level transaction undo/releae other than proced…

### PR DESCRIPTION
**Pull request by jxchinaren**
…ure level; Add various test cases to stress the code;

### CODE REVIEW

File: tests/frontend/org/voltdb/regressionsuites/TestCatchExceptionsInProcedure.java [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (jxchinaren) [inline tests/frontend/org/voltdb/regressionsuites/TestCatchExceptionsInProcedure.java]: Minor, the automatic license checker tool adds duplicated license information here, needs to clean it up.

File: src/frontend/org/voltdb/ProcedureRunner.java [hunk 2]
```diff
@@ -144,6 +144,9 @@
     //  - track the current call to voltExecuteSQL for logging progress
     protected int m_batchIndex;
 
+    /** boolean flag to mark whether the previous batch execution has EE exception or not.*/
+    private long m_spBigBatchBeginToken = -100L;
```
Comment (nshi) [inline src/frontend/org/voltdb/ProcedureRunner.java]: What's the magic -100?

File: src/frontend/org/voltdb/ProcedureRunner.java [hunk 3]
```diff
@@ -703,6 +698,13 @@ public void voltQueueSQL(final String sql, Object... args) {
             }
             m_seenFinalBatch = isFinalSQL;
 
+            // should check whether the batch is read only or not
+            // e.g. read only query may have timed out...
+            if (!m_isSinglePartition && m_txnState.needsRollback()) {
+                throw new RuntimeException("Multi-partition procedure " + m_procedureName +
```
Comment (nshi) [inline src/frontend/org/voltdb/ProcedureRunner.java]: Consider throwing a more concrete exception like `VoltAbortException`.

File: src/frontend/org/voltdb/ProcedureRunner.java [hunk 4]
```diff
@@ -715,6 +717,10 @@ public void voltQueueSQL(final String sql, Object... args) {
             }
             // otherwise, break it into sub-batches
             else {
+                if (! m_isReadOnly) {
+                    // increase 1 here to mark the next executing undo token
+                    m_spBigBatchBeginToken = m_site.getLatestUndoToken() + 1;
```
Comment (nshi) [inline src/frontend/org/voltdb/ProcedureRunner.java]: You have to make sure that `m_spBigBatchBeginToken` is reset at the end of the transaction or the next transaction of the same procedure will have a positive `m_spBigBatchBeginToken`, causing incorrect undo.

File: src/frontend/org/voltdb/dtxn/TransactionState.java [hunk 5]
```diff
@@ -142,9 +142,9 @@ public long getBeginUndoToken()
     }
 
     // Assume that rollback-ness is a latch.
```
Comment (nshi) [inline src/frontend/org/voltdb/dtxn/TransactionState.java]: Comment out-of-date. It's no longer a latch now.

File: src/frontend/org/voltdb/iv2/Site.java [hunk 6]
```diff
@@ -129,7 +129,7 @@
     final SiteTaskerQueue m_scheduler;
 
     /*
-     * There is really no legit reason to touch the initiator mailbox from the site,
```
Comment (nshi) [inline src/frontend/org/voltdb/iv2/Site.java]: Legit is a legit word, but I guess we should use more formal form in writing.

File: tests/frontend/org/voltdb/regressionsuites/TestCatchExceptionsInProcedure.java [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+        for (int noDup: noDuplicates) {
+            // exception in the first 200 batch
+            bigBatchChecker(client, 0, noDup, 0, 0, new double[]{10.1}, BIGBATCHTESTSIZE);
+            bigBatchChecker(client, 0, noDup, 1, 0, new double[]{10.1, 500.1}, BIGBATCHTESTSIZE+1);
+            bigBatchChecker(client, 0, noDup, 1, 1, new double[]{}, 0);
+            bigBatchChecker(client, 1, noDup, 0, 0, new double[]{prev, 10.1}, BIGBATCHTESTSIZE+1);
+            bigBatchChecker(client, 1, noDup, 1, 0, new double[]{prev, 10.1, 500.1}, BIGBATCHTESTSIZE+2);
+            bigBatchChecker(client, 1, noDup, 1, 1, new double[]{}, 0);
+        }
+    }
```
Comment (nshi) [inline tests/frontend/org/voltdb/regressionsuites/TestCatchExceptionsInProcedure.java]: It's also worth adding a test to check the following scenario,

- Call procedure A with big batches, rollback in the middle of a big batch
- Call procedure A with *no* big batch, rollback the whole transaction
- Make sure data is as expected in the database

**Comment by nshi** [comment]
Review completed.

File: src/frontend/org/voltdb/iv2/Site.java (hunk 6 above)
Comment (jxchinaren) [inline src/frontend/org/voltdb/iv2/Site.java]: Cited from Wiki: "Legit is a slang abbreviation of legitimate"

Good to learn a new word "legit" here.

**Comment by jxchinaren** [comment]
I get the confirmation to merge from Ning. I will close this pull request and squash merge this work. Thanks.

**Pull request closed without merging**